### PR TITLE
[WIP] Implement ExUnit max-failures feature

### PR DESCRIPTION
**Pull request by eksperimental**
Implements: #4586
Related: #7480, #8129

Initially it was proposed to call this option "max-fail", but the terminology used in ExUnit is failures,
so I went with "max-failures"

This is a work in progress, I still need to add some more tests related to concurrent tests, but I think it's in a good stage to share it and get feedback.

### CODE REVIEW

**Comment by eksperimental** [comment]
Please if you have a real life scenario where to use it,
it's fully functional (or at least supposed to), and give me some feedback and the logic of this
was more complex than I thought and there may be yet some corner-cases that haven't been covered

**Comment by Kartstig** [comment]
Code looks great. Thank you for working on this!

File: lib/mix/lib/mix/tasks/test.ex [hunk 1]
```diff
@@ -288,6 +290,7 @@ defmodule Mix.Tasks.Test do
     cover: :boolean,
     trace: :boolean,
     max_cases: :integer,
+    max_failures: :string,
```
Comment (fertapric) [inline lib/mix/lib/mix/tasks/test.ex]: The `--slowest` flag uses `0` instead of `:infinity` to disable the feature, what do you think about taking the same approach here?

Comment (eksperimental) [reply lib/mix/lib/mix/tasks/test.ex]: @fertapric you have to remember that you will always have one more failure than max-failure.
so `--max-failure 0`  will stop on the first failure.
If we want something like that we should rename the feature to something like "stop-on-nth-failure" (I'm not really good at coming up with new names :-)

Comment (hylkealons) [reply lib/mix/lib/mix/tasks/test.ex]: agree with @fertapric, it makes more sense to have `--max-failures 1` -> stopping after first failure.

**Comment by hylkealons** [comment]
Great work @eksperimental, I will try to test it and give some feedback. Can't wait to see this merged 😎

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (eksperimental) [reply lib/mix/lib/mix/tasks/test.ex]: I have changed the behaviour to stop when max-failures is reached. Though I haven't changed the default value being `:infinity` until we reach consensus.

**Comment by eksperimental** [comment]
I have added some new commits, the main change is that before we were executing setup and setup_all callbacks every time, but now we follow the rule that if the test is discarded or excluded, the callbacks are never run. So if we have reached the max-failures limit, we don't run any callback

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (fertapric) [reply lib/mix/lib/mix/tasks/test.ex]: I thought about that too, and I also think `:infinity` is better for this feature. Using `0` with `--slowest` makes sense, but using it for disabling `--max-failures` would not be semantically correct. 

However, the behavior of `--max-failures 0` is a bit confusing. I guess I would expect to not run any tests 🤔

Comment (eksperimental) [reply lib/mix/lib/mix/tasks/test.ex]: I think we could raise if max-failures is 0, since it's probably an undesired behaviour

File: lib/ex_unit/lib/ex_unit/on_exit_handler.ex [hunk 2]
```diff
@@ -2,20 +2,40 @@ defmodule ExUnit.OnExitHandler do
...
 
   @spec start_link(keyword()) :: {:ok, pid}
   def start_link(_opts) do
     Agent.start_link(fn -> :ets.new(@name, @ets_opts) end, name: @name)
   end
 
-  @spec register(pid) :: :ok
-  def register(pid) when is_pid(pid) do
-    :ets.insert(@name, {pid, nil, []})
+  @spec register(pid, term) :: :ok
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: Is the `manager` also here `{pid, pid}`? (same for `get_registered_pids /1`)

Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: yes. I guess we could enforce that in the spec

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 3]
```diff
@@ -323,18 +440,31 @@ defmodule ExUnit.Runner do
...
+          %{test_or_case | state: state}
+      end
+    else
+      test_or_case
     end
   end
 
   ## Helpers
 
+  defp max_failures_reached?(failure_counter, max_failures)
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/runner.ex]: What do you think about pattern-match on `:infinity` to speed up the happy case?

```elixir
defp max_failures_reached?(_, :infinity), do: false

defp max_failures_reached?(manager, max_failures) when is_tuple(manager) do
  get_failure_counter(manager) >= max_failures
end

defp max_failures_reached?(failures, max_failures), do: failures >= max_failures
```

File: lib/ex_unit/lib/ex_unit/on_exit_handler.ex (hunk 2 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: I have added it as a type

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 3 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: I'm not too sure about removing the type checking guards since we are not validating max_failures when the functions is called directly.

File: lib/mix/lib/mix/tasks/test.ex [hunk 4]
```diff
@@ -569,6 +575,34 @@ defmodule Mix.Tasks.Test do
     end
   end
 
+  defp validate_max_failures_opts(opts) do
```
Comment (fertapric) [inline lib/mix/lib/mix/tasks/test.ex]: What do you think about using `Integer.parse/2` to avoid control-flow via exceptions?

```elixir
defp validate_max_failures_opts(opts) do
  case Keyword.fetch(opts, :max_failures) do
    {:ok, value} when value not in ["infinity", ":infinity"] ->
      case Integer.parse(value) do
        {max_failures, _} when max_failures >= 1 -> Keyword.put(opts, :max_failures, :infinity)
        :error -> raise "--max-failuers : Expected type positive integer or infinity, got '#{value}'"
      end
      
    _ -> 
      Keyword.put(opts, :max_failures, :infinity)
  end
end
```

Additionally, it covers:

* the edge case of `--max-failures 0`
* invalid values like `--max-failures invalid`. Similar to what other flags behave:

    ```
    $ mix test --slowest a
    ** (Mix) Could not invoke task "test": 1 error found!
    --slowest : Expected type integer, got "a"
    ```

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 3 above)
Comment (fertapric) [reply lib/ex_unit/lib/ex_unit/runner.ex]: That's also part of another comment [I was currently writing](https://github.com/elixir-lang/elixir/pull/8174/files#r217171086). If applied, it would ensure that `max_failures` is either `:infinity` or a positive integer 🙂

File: lib/ex_unit/lib/ex_unit.ex [hunk 5]
```diff
@@ -65,12 +65,18 @@ defmodule ExUnit do
     1. Passed (also represented by `nil`)
     2. Failed
     3. Skipped (via @tag :skip)
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit.ex]: Minor nitpick: do you mind adding backticks here? `@tag :skip`, `:exclude`, `setup_all/1,2` 🙂

File: lib/mix/lib/mix/tasks/test.ex (hunk 4 above)
Comment (eksperimental) [reply lib/mix/lib/mix/tasks/test.ex]: looks greate. i renamed it `normalize_max_failures`, and I changed a bit the logic of your code to deal with invalid decimal numbers
$ mix test --max-failures 0
** (RuntimeError) --max-failures : Expected type positive integer or infinity, got '0'

File: lib/mix/lib/mix/tasks/test.ex [hunk 6]
```diff
@@ -569,6 +575,22 @@ defmodule Mix.Tasks.Test do
     end
   end
 
+  defp normalize_max_failures(opts) do
+    case Keyword.fetch(opts, :max_failures) do
+      value when value in ["infinity", ":infinity"] ->
```
Comment (fertapric) [inline lib/mix/lib/mix/tasks/test.ex]: I think this should be `{:ok, value}`

Comment (fertapric) [reply lib/mix/lib/mix/tasks/test.ex]: 🤔And I think the `:error` value cannot be matched

Comment (eksperimental) [reply lib/mix/lib/mix/tasks/test.ex]: oops

File: lib/ex_unit/lib/ex_unit/on_exit_handler.ex [hunk 7]
```diff
@@ -2,20 +2,42 @@ defmodule ExUnit.OnExitHandler do
...
   @ets_opts [:public, :named_table, read_concurrency: true, write_concurrency: true]
 
+  # ETS column numbers
+  @pid 1
+  @supervisor 3
+  @on_exit 4
+
   use Agent
 
+  @type manager :: ExUnit.EventManager.use(Agent)
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: Should this be `ExUnit.EventManager.manager()`?

Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: fixed. thank you for all your input

File: lib/mix/lib/mix/tasks/test.ex [hunk 8]
```diff
@@ -569,6 +575,22 @@ defmodule Mix.Tasks.Test do
     end
   end
 
+  defp normalize_max_failures(opts) do
+    case Keyword.fetch(opts, :max_failures) do
```
Comment (fertapric) [inline lib/mix/lib/mix/tasks/test.ex]: 😅 I think you missed the comment related to the `:error`. I think this code will raise `** (CaseClauseError) no case clause matching: :error`

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 9]
```diff
@@ -299,14 +351,79 @@ defmodule ExUnit.Runner do
...
+    end
+  end
+
+  defp process_failure(%ExUnit.Test{} = test, _options) do
+    {:ok, test}
+  end
+
+  # Stops all running tests in the suite
+  defp terminate_all_running_processes(manager, parent_pid, current_pid) do
+    for registered_pid when registered_pid not in [parent_pid, current_pid] <-
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/runner.ex]: Minor change for readability:

```elixir
for registered_pid <- ExUnit.OnExitHandler.get_registered_pids(manager),
    registered_pid not in [parent_pid, current_pid],
    Process.alive?(registered_pid) do
  Process.exit(registered_pid, :max_failures_reached)
end
```

File: lib/mix/lib/mix/tasks/test.ex (hunk 8 above)
Comment (fertapric) [reply lib/mix/lib/mix/tasks/test.ex]: I think it should work now 👍

File: lib/ex_unit/lib/ex_unit/runner.ex (hunk 9 above)
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: proper use of filters. thank you

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 10]
```diff
@@ -244,48 +263,81 @@ defmodule ExUnit.Runner do
...
+        :timer.tc(fn ->
+          if max_failures_reached?(config.manager, config.max_failures) do
+            %{test | state: {:not_executed, @not_executed_msg}, time: 0}
+          else
             case exec_test_setup(test, context) do
               {:ok, test} ->
                 exec_test(test)
 
               {:error, test} ->
+                # %{test | state: {:invalid, test}}
```
[the remark this answers is not in the document]
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: no.  I need to investigate why setting it as :invalid it will not work

[the remark this answers is not in the document]
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/runner.ex]: i mean, it will not pass the tests

File: lib/ex_unit/lib/ex_unit/runner.ex [hunk 11]
```diff
@@ -299,14 +351,78 @@ defmodule ExUnit.Runner do
...
+        registered_pid not in [parent_pid, current_pid],
+        Process.alive?(registered_pid) do
+      Process.exit(registered_pid, :max_failures_reached)
+    end
+  end
+
+  defp get_failure_counter(manager),
+    do: ExUnit.OnExitHandler.get_failure_counter(manager)
+
+  defp increment_failure_counter(manager, test, increment \\ 1)
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/runner.ex]: [`ExUnit.OnExitHandler.increment_failure_counter/2`](https://github.com/elixir-lang/elixir/pull/8174/files#diff-258d820db9fe3b2672410c8255211706R88) also validates the `increment` argument, so we might not need to also check for it here.

This internal function is only called from [here](https://github.com/elixir-lang/elixir/pull/8174/files#diff-1e5179d66aaabef6fbc8efa93ae33493R368) and [here](https://github.com/elixir-lang/elixir/pull/8174/files#diff-1e5179d66aaabef6fbc8efa93ae33493R387). Keeping that in mind,  I think the last clause  `defp increment_failure_counter(manager, _, _increment)` won't be executed; and the `tag` on `ExUnit.TestModule` won't never be `:not_executed` 🙂

I think everything could be simplified to (the `tag` is checked on the parent calls):

```elixir
defp increment_failure_counter(manager, increment \\ 1) do
  ExUnit.OnExitHandler.increment_failure_counter(manager, increment)
end
```

File: lib/ex_unit/lib/ex_unit/on_exit_handler.ex [hunk 12]
```diff
@@ -51,11 +73,29 @@ defmodule ExUnit.OnExitHandler do
...
   @spec run(pid, timeout) :: :ok | {Exception.kind(), term, Exception.stacktrace()}
   def run(pid, timeout) when is_pid(pid) do
-    [{^pid, sup, callbacks}] = :ets.take(@name, pid)
+    [{^pid, _manager, sup, callbacks}] = :ets.take(@name, pid)
     error = terminate_supervisor(sup, timeout)
     exec_on_exit_callbacks(Enum.reverse(callbacks), timeout, error)
   end
 
+  @spec start_failure_counter(manager) :: :ok
+  def start_failure_counter(manager) do
```
Comment (fertapric) [inline lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: Does the "failure counter" belong to the `ExUnit.OnExitHandler`, conceptually? (just to double check 🙂)

Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: It doesn't . I thought of that too. But thought creating another ETS for just a counter would be overkill.
let's ask what @josevalim  thinks about it

Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/on_exit_handler.ex]: probably we could implement it using an Agent

File: lib/ex_unit/lib/ex_unit/event_manager.ex [hunk 13]
```diff
@@ -54,8 +57,8 @@ defmodule ExUnit.EventManager do
 
   def module_started(ref, test_module) do
     # TODO: Remove case_started in Elixir v2.0
-    notify(ref, {:module_started, test_module})
     notify(ref, {:case_started, Map.put(test_module, :__struct__, ExUnit.TestCase)})
+    notify(ref, {:module_started, test_module})
```
[the remark this answers is not in the document]
Comment (fertapric) [reply lib/ex_unit/lib/ex_unit/event_manager.ex]: I wondered the same. I think it’s for the code to be closer to the comment. And for what I saw, I think the order should not matter :)

**Comment by josevalim** [comment]
Thanks @eksperimental.

I do have some concerns about the implementation. I would like to not have to introduce a new status, not_executed. I believe we can abort the runner immediately and just not care about any subsequent test.

The other thing is that I don't like the runner is using the `ExUnit.OnExitHandler` process to retrieve information about current processes. The runner should have all of this information. The only reason the ExUnit.OnExitHandler process exists is because the test needs to send information to the runner.

Doing those changes should simplify the implementation a lot.

**Comment by josevalim** [comment]
@eksperimental if the user is interested in quick feedback, then I would propose for us to abort as soon as possible. There is no reason to wait for the rest of the test suite to finish loading only to not run any of it. It will be less LOC in the implementation too. :)

**Comment by eksperimental** [comment]
Well i guess we could figure out that number based on the total tests, the failed ones, but we would need to keep track of the passed tests which we are not doing now

**Comment by eksperimental** [comment]
So my approach is to create a exunit.runner.tracker module that will keep track of the test pids and thenl counter in a ets table, similar to how was done in onexit.handler. 
Are you ok with this, or you prefer something different?

File: lib/ex_unit/lib/ex_unit/event_manager.ex (hunk 13 above)
[the remark this answers is not in the document]
Comment (eksperimental) [reply lib/ex_unit/lib/ex_unit/event_manager.ex]: Yes, as fertapric said. The comment is right above the line that needs to be removed

**Comment by eksperimental** [comment]
Or just an Agent

**Comment by josevalim** [comment]
> So my approach is to create a exunit.runner.tracker module that will keep track of the test pids and the failure counter in a ets table, similar to how was done in onexit.handler.

IMO the existing runner should know all of the pids and should be able to track all information about the processes it runs. We should not need a separate state storage. Everything could be kept in the runner state. Maybe by adding one or two new fields.
